packages/ui/src/components/TagList/index.tsx

```diff
@@ -82,77 +107,197 @@ export const TagList = ({
   className,
   'data-testid': dataTestId,
 }: TagListProps) => {
-  let tmpThreshold = threshold
-  if (
-    tags.length > 0 &&
-    tags
-      .slice(0, tmpThreshold)
-      .reduce<string>((acc, tag) => acc + getTagLabel(tag), '').length >
-      maxLength
-  ) {
-    // If total tags length in characters is above maxLength,
-    // threshold is decremented in order to prevent too many long tags displayed.
-    tmpThreshold -= 1
-  }
-  const hasManyTags = tags.length > tmpThreshold || false
-  const visibleTagsCount = hasManyTags ? tmpThreshold : tags.length
+  const containerRef = useRef<HTMLDivElement>(null)
+  const measureRef = useRef<HTMLDivElement>(null)
+  const popoverTriggerRef = useRef<HTMLDivElement>(null)
+
+  const [isPopoverVisible, setIsPopoverVisible] = useState(false)
+  const [popoverTriggerWidth, setPopoverTriggerWidth] = useState(0)
+  const [visibleTags, setVisibleTags] = useState<TagType[]>([])
+  const [hiddenTags, setHiddenTags] = useState<TagType[]>([])
+
+  // Compute tmpThreshold, potentially visible tags and surely hidden tags
+  const memoizedResult = useMemo(() => {
+    let tmpThreshold = threshold
+    if (
+      tags.length > 0 &&
+      tags
+        .slice(0, tmpThreshold)
+        .reduce<string>((acc, tag) => acc + getTagLabel(tag), '').length >
+        maxLength
+    ) {
+      tmpThreshold -= 1
+    }
+
+    const potentiallyVisibleTagsLength =
+      tags.length > tmpThreshold || false ? tmpThreshold : tags.length
+    const potentiallyVisibleTags = tags.slice(0, potentiallyVisibleTagsLength)
+    const surelyHiddenTags = tags.slice(potentiallyVisibleTagsLength)
+
+    return {
+      tmpThreshold,
+      potentiallyVisibleTags,
+      surelyHiddenTags,
+    }
+  }, [maxLength, tags, threshold])
+
+  const { tmpThreshold, potentiallyVisibleTags, surelyHiddenTags } =
+    memoizedResult
+
+  // compute visible tags and hidden ones based on the container width and
+  // what can fit into it from the potentially visible tags
+  useEffect(() => {
+    if (!tags.length || !containerRef.current || !measureRef.current) {
+      return
+    }
+
+    if (multiline) {
+      setVisibleTags(potentiallyVisibleTags)
+      setHiddenTags(surelyHiddenTags)
+
+      return
+    }
+
+    const parentWidth = containerRef.current.parentElement?.clientWidth || 0
+    let accumulatedWidth = 0
+    const measuredVisibleTags = []
+    const measuredHiddenTags = []
+
+    const measureElements = measureRef.current.children[0].children
+    for (let i = 0; i < measureElements.length; i += 1) {
+      accumulatedWidth +=
+        (measureElements[i] as HTMLDivElement).offsetWidth +
+        parseInt(TAGS_GAP, 10)
+      if (accumulatedWidth <= parentWidth) {
+        measuredVisibleTags.push(tags[i])
+      } else {
+        measuredHiddenTags.push(tags[i])
+      }
+    }
```

Comment:
We tend to avoid `for` / `let` and using `.push()` on arrays due to mutation. Instead you could do it with a reduce:

```tsx
const { measuredVisibleTags, measuredHiddenTags, accumulatedWidth } = measureElements.reduce((acc, element, index) => {
  const newAccumulatedWidth = acc.accumulatedWidth + (measureElements[i] as HTMLDivElement).offsetWidth +
        parseInt(TAGS_GAP, 10)
  
  return 
     ({ 
       measuredVisibleTags: [...(acc.measuredVisibleTags), accumulatedWidth <= parentWidth && tags[i]], 
       measuredHiddenTags: [...(acc.measuredHiddenTags), accumulatedWidth > parentWidth && tags[i]], 
       accumulatedWidth: newAccumulatedWidth 
      })

}, {measuredVisibleTags: [], measuredHiddenTags: [], accumulatedWidth: 0})
```